File: sdk/go/name_data_test.go
package bunqueue

import (
	"reflect"
	"testing"
)

func TestNameDataWireSeparation(t *testing.T) {
	connection, socket := scriptedConnection(t)
	queue := NewQueueWithConnection("named", connection)
	data := map[string]any{"name": "customer-visible", "to": "a@b.c"}

	job, err := queue.Add("send-email", data, nil)
	if err != nil {
		t.Fatal(err)
	}
	command := decodeWrittenCommand(t, socket)
	if command["name"] != "send-email" || !reflect.DeepEqual(command["data"], data) {
		t.Fatalf("PUSH name/data were not separate: %#v", command)
	}
	if job.Name() != "send-email" || !reflect.DeepEqual(job.Data(), data) {
		t.Fatalf("producer stub lost name/data: name=%q data=%#v", job.Name(), job.Data())
	}
}

func TestBulkAndJobReadersPreserveModernAndLegacyPayloads(t *testing.T) {
	connection, socket := scriptedConnection(t)
	queue := NewQueueWithConnection("named-bulk", connection)
	_, err := queue.AddBulk([]BulkEntry{
		{Name: "object-job", Data: map[string]any{"name": "user-name", "value": 1}},
		{Name: "scalar-job", Data: false},
	})
	if err != nil {
		t.Fatal(err)
	}
	command := decodeWrittenCommand(t, socket)
	jobs := asSlice(command["jobs"])
	if asMap(jobs[0])["name"] != "object-job" || asMap(jobs[1])["data"] != false {
		t.Fatalf("PUSHB name/data were not separate: %#v", jobs)
	}

	modern := newJob(map[string]any{
		"name": "modern-op", "data": map[string]any{"name": "user-name", "value": 1},
	}, connection, "")
	legacy := newJob(map[string]any{
		"data": map[string]any{"name": "legacy-op", "value": 2},
	}, connection, "")
	scalar := newJob(map[string]any{"name": "scalar-op", "data": 42}, connection, "")

	if modern.Name() != "modern-op" || !reflect.DeepEqual(modern.Data(), modern.Raw["data"]) {
		t.Fatalf("modern job changed user data: name=%q data=%#v", modern.Name(), modern.Data())
	}
	if legacy.Name() != "legacy-op" || !reflect.DeepEqual(legacy.Data(), map[string]any{"value": 2}) {
		t.Fatalf("legacy envelope was not decoded: name=%q data=%#v", legacy.Name(), legacy.Data())
	}
	if scalar.Name() != "scalar-op" || !reflect.DeepEqual(scalar.Data(), 42) {
		t.Fatalf("scalar payload was not preserved: name=%q data=%#v", scalar.Name(), scalar.Data())
	}
}

File: sdk/typescript/tests/name-data.test.ts
import { describe, expect, test } from 'bun:test';
import type { Connection } from '../src/connection.js';
import { Job } from '../src/job.js';
import { Queue } from '../src/queue.js';

class RecordingConnection {
  readonly commands: Array<Record<string, unknown>> = [];

  async call(command: Record<string, unknown>): Promise<Record<string, unknown>> {
    this.commands.push(command);
    return command.cmd === 'PUSHB'
      ? { ok: true, ids: ['bulk-a', 'bulk-b'] }
      : { ok: true, id: 'single' };
  }
}

describe('name and data wire separation', () => {
  test('PUSH and producer stubs keep an object data.name as user data', async () => {
    const connection = new RecordingConnection();
    const queue = new Queue('named', { connection: connection as unknown as Connection });
    const data = { name: 'customer-visible', to: 'a@b.c' };

    const job = await queue.add('send-email', data);

    expect(connection.commands[0]).toEqual({
      cmd: 'PUSH',
      queue: 'named',
      name: 'send-email',
      data,
    });
    expect(job.name).toBe('send-email');
    expect(job.data).toEqual(data);
  });

  test('PUSHB preserves names and primitive payloads without wrapping', async () => {
    const connection = new RecordingConnection();
    const queue = new Queue<unknown>('named-bulk', {
      connection: connection as unknown as Connection,
    });

    const jobs = await queue.addBulk([
      { name: 'object-job', data: { name: 'user-name', value: 1 } },
      { name: 'scalar-job', data: 42 },
    ]);

    expect(connection.commands[0]).toEqual({
      cmd: 'PUSHB',
      queue: 'named-bulk',
      jobs: [
        { name: 'object-job', data: { name: 'user-name', value: 1 } },
        { name: 'scalar-job', data: 42 },
      ],
    });
    expect(jobs.map((job) => [job.name, job.data])).toEqual([
      ['object-job', { name: 'user-name', value: 1 }],
      ['scalar-job', 42],
    ]);
  });

  test('Job prefers top-level name and only unwraps a legacy data envelope', () => {
    const modern = new Job({ name: 'modern-op', data: { name: 'user-name', value: 1 } });
    const legacy = new Job({ data: { name: 'legacy-op', value: 2 } });
    const scalar = new Job({ name: 'scalar-op', data: false });

    expect([modern.name, modern.data]).toEqual(['modern-op', { name: 'user-name', value: 1 }]);
    expect([legacy.name, legacy.data]).toEqual(['legacy-op', { value: 2 }]);
    expect([scalar.name, scalar.data]).toEqual(['scalar-op', false]);
  });

  test('worker-owned Job mutations forward the delivery token', async () => {
    const connection = new RecordingConnection();
    const job = new Job(
      { id: 'leased-job', queue: 'leased' },
      connection as unknown as Connection,
      'lease-token'
    );

    await job.retry();
    await job.changeDelay(30_000);
    await job.moveToDelayed(60_000);
    await job.discard();

    expect(connection.commands).toEqual([
      { cmd: 'MoveToWait', id: 'leased-job', token: 'lease-token' },
      { cmd: 'ChangeDelay', id: 'leased-job', delay: 30_000, token: 'lease-token' },
      { cmd: 'MoveToDelayed', id: 'leased-job', delay: 60_000, token: 'lease-token' },
      { cmd: 'Discard', id: 'leased-job', token: 'lease-token' },
    ]);
  });
});
